### PR TITLE
finished plugin

### PR DESCRIPTION
**Pull request by maciekpaprocki**
- Removed herbert 
- Added old items rewrite
- Improved architecture
- Fixed exception
- Improved speed

### CODE REVIEW

File: src/Action/Post.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
+<?php
+/**
+ * User: maciej
```
Comment (ecoad) [inline src/Action/Post.php]: Do remove these phpdoc comments. Ta

File: src/Action/Post.php [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+class Post {
+
+	/**
+	 * Delete Redirect from Tapestry Redis store when redirects are deleted
+	 *
+	 * @param int $post_id The post ID.
+	 * @param post $post The post object.
+	 * @param bool $update Whether this is an existing post being updated or not.
+	 */
+	public static function delete_post( $post_id ) {
```
Comment (ecoad) [inline src/Action/Post.php]: The params here doesn't match the documentation above

File: src/Action/Post.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+
+	/**
+	 * Update Tapestry Redis store when Redirects are saved
+	 *
+	 * @param int $post_id The post ID.
+	 * @param \WP_Post $post The post object.
+	 * @param bool $update Whether this is an existing post being updated or not.
+	 */
+	public static function save_post( $post_id, $post ) {
```
Comment (ecoad) [inline src/Action/Post.php]: The params here do not match the documentation above

File: src/Driver/Csv.php [hunk 4]
```diff
@@ -0,0 +1,81 @@
+<?php
+/**
+ * User: maciej
```
Comment (ecoad) [inline src/Driver/Csv.php]: Remove

File: src/Driver/DriverInterface.php [hunk 5]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+namespace ShortlistMedia\WpBigRedirector\Driver;
```
Comment (ecoad) [inline src/Driver/DriverInterface.php]: Missing the `Friendly` part from the namespace. Also we always add `Plugin` as a suffix. So namespace should be: `ShortlistMedia\WpBigFriendlyRedirectorPlugin\Driver`.
We don't normally add the `ShortlistMedia` but but we probably should 👍

File: src/Driver/Wordpress.php [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	/**
+	 * @return array
+	 */
+	public function getAll() {
+
+		global $wpdb;
+
+		$res = $wpdb->get_results( "SELECT m1.meta_value as `from`, m2.meta_value as `to`
```
Comment (ecoad) [inline src/Driver/Wordpress.php]: Please add a comment what this chunk of code is doing

File: src/Driver/Wordpress.php [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+						JOIN {$wpdb->postmeta} m1 ON p.ID = m1.post_id 
+						JOIN {$wpdb->postmeta} m2 ON p.ID = m2.post_id 
+						WHERE m1.meta_key = '_redirect_rule_from' 
+							AND m2.meta_key = '_redirect_rule_to' 
+							AND p.post_status = 'publish' 
+							AND p.post_type = 'redirect_rule'", ARRAY_N );
+
+
+		$post_types = $this->getVisiblePostTypes();
+		$preRes     = $wpdb->get_results( "SELECT p.guid, p.ID
```
Comment (ecoad) [inline src/Driver/Wordpress.php]: Please add a comment what this chunk of code is doing

File: src/Driver/Wordpress.php [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+						FROM {$wpdb->posts} p
+							WHERE p.post_status = 'publish' 
+							AND p.post_type IN($post_types)", ARRAY_N );
+
+
+		foreach ( $preRes as $pre_re ) {
```
Comment (ecoad) [inline src/Driver/Wordpress.php]: Please add a comment what this chunk of code is doing

File: src/Driver/Wordpress.php [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+		$ret = [];
+		foreach ( $res as $val ) {
+			$ret[ $val[0] ] = $val[1];
+		}
+
+		return $ret;
+	}
+
+	public function getVisiblePostTypes() {
```
Comment (ecoad) [inline src/Driver/Wordpress.php]: Please add a function-level comment and comment any chunks of code that are not immediately obvious.

File: src/Repository/Redirect.php [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+	 *  Updates sources using csv
+	 */
+	public function updateFromCsv() {
+		$csvDriver = new Csv();
+		$this->getDriver()->setFromArray( $csvDriver->getAll() );
+	}
+
+	public function updateFromWordpressRedirects() {
+		$wordpressRedirectsDriver = new Wordpress();
+		$this->getDriver()->setFromArray( $wordpressRedirectsDriver->getAll() );
```
Comment (ecoad) [inline src/Repository/Redirect.php]: How big can these arrays be before the memory is too large to run on a 128MB process?

File: src/cli.php [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+ * ## EXAMPLES
+ *
+ *     # Add all a specified post
+ *     wp bfr update
+ *
+ */
+
+\WP_CLI::add_command( 'bfr update', function () {
+
+	\Symfony\Component\Debug\Debug::enable();
```
Comment (ecoad) [inline src/cli.php]: Is this still needed now Herbert has been removed?

File: src/Repository/Redirect.php (hunk 10 above)
Comment (maciekpaprocki) [inline src/Repository/Redirect.php]: it's running through console mostly so it should have more memory but either way, that should last for 500k+ items(i think like 14k currently).

File: src/cli.php (hunk 11 above)
Comment (maciekpaprocki) [inline src/cli.php]: Unfortunately yes, herbert is included still in other plugins.

**Comment by maciekpaprocki** [comment]
@ecoad

**Pull request merged**